proposals/user-scripts-api.md

```diff
@@ -0,0 +1,189 @@
+  callback?: function
+)
+browser.userScripts.unregister(
+  filter?: UserScriptFilter[],
+  callback?: function
+)
+browser.userScripts.getScripts(
+  filter?: UserScriptFilter[],
+  callback?: function
+)
+browser.userScripts.update(
+  scripts?: RegisteredUserScript[],
+  callback?: function
+)
+```
+
+### Requirements
+
+#### A. A mechanism to execute code in the main world
+
+- User scripts can be registered in the `MAIN` world.
+- Extension can customize the `USER_SCRIPT` world's CSP to inject a script tag into the host page (expanded in requirement B)
+
+#### B. The ability to execute code (with a separate CSP) in a world different from the main world and the extension's isolated world
+
+##### `USER_SCRIPT` World
+
+- Exempt from the page's CSP, can customize its own CSP (e.g can allow injection of `unsafe-inline`).
+- Can communicate with the extension using extension messaging APIs (expanded under [Messaging](#Messaging)).
+- Isolated from the web page (similar to other isolated worlds), but will be largely un-permissioned. It will not have access to any extension APIs, except messaging APIs mentioned, and will not have any cross-origin exceptions (these don't exist in Chrome, but do in other browsers).  In the future, we may bring more APIs (such as chrome.dom APIs or a dedicated API to execute a script in the main world), but these APIs will not grant any additional privilege to the script beyond what it has to access the page content.
+- Can communicate with different JS worlds via `window.postMessage()`. A dedicated API to communicate between worlds is being considered as part of future work.
+- Shares DOM with the web page.  Code from both worlds cannot directly interact with each other, except through DOM APIs.
+- When an asymmetric security relationship may exist, the `MAIN` world is considered to be less privileged than the `USER_SCRIPT` world
+
+##### Configuration
+
+```
+browser.userScripts.configureWorld(
+  csp?: string
+  enableMessaging?: bool
+)
+```
+where
+- If csp is defined, it is used in the `USER_SCRIPT` world. Otherwise, the `ISOLATED` world CSP is used.
+- If `enableMessaging` is true, messaging APIs are exposed. Otherwise, if false or undefined, messaging APIs are not exposed.
+In the future, if we allow multiple user script worlds (see section in Future Work below), this method can be expanded to allow for a user script world identifier to customize a single user script world.
+
+##### Messaging
+
+User scripts can send messages to the extension using extension messaging APIs: `browser.runtime.sendMessage()` and `browser.runtime.connect()`. We leverage the runtime API (instead of introducing new userScripts.onMessage- and userScripts.sendMessage-style values) in order to keep extension messaging in the same API.  There is precedent in this (using the same API namespace to send messages from a different (and less trusted) context, as chrome.runtime is also the API used to send messages from web pages.
```

Comment:
I think that it makes more sense to add a new namespace for messaging with the `userScripts` API, instead of `runtime.sendMessage`/`connect`, for the following reasons:
- If future work introduces additional functionality, it would need a namespace to hold the methods. We may as well introduce the namespace now, e.g. `userScriptContent`.
- In this API, eventually code will be running in the `MAIN` world and the `USER_SCRIPTS` world. It may be confusing if `runtime.sendMessage`/`connect` from `MAIN` triggers `runtime.onMessageExternal`/`runtime.onConnectExternal`, while exactly the same code triggers `runtime.onUserScriptMessage`/`runtime.onUserScriptConnect` in `USER_SCRIPT` (... and `runtime.onMessage`/`runtime.onConnect` in `ISOLATED_WORLD` through content scripts, e.g. the `scripting` API).
  - I already see frequent confusion among developers who do not understand the difference between the `runtime.sendMessage`/`connect` APIs in web pages vs content scripts, where the observable difference is that the former requires an `extensionId` parameter.
- The `extensionId` parameter allows communication with other extensions. Since the method is explicitly meant to be used by user script managers, I don't think that this would be desirable. This could easily be addressed at the implementation level by rejecting the `extensionId` parameter, but good luck with documenting that...

What do y'all think of these APIs instead:

- `browser.userScriptContent.sendMessage` / `connect`
- `browser.userScripts.onMessage` / `onConnect`